tests/frontend/requestValidator.js

```diff
@@ -0,0 +1,64 @@
+const EventEmitter = require("events");
+const { sleep } = require("../utils");
+const { expect } = require("@playwright/test");
+const { uid } = require("uid");
+
+const eventEmitter = new EventEmitter();
+
+const dispatch = (type) => {
+  eventEmitter.emit(type);
+};
+
+const subscribe = (type, callback) => {
+  eventEmitter.on(type, callback);
+};
+
+const unsubscribe = (type, callback) => {
+  eventEmitter.removeListener(type, callback);
+};
+
+const wrapWithRequestId = (func) => {
+  return () => {
+    const requestId = uid();
+    func(requestId);
+    return requestId;
+  };
+};
+
+const execAndWaitForRequest = async (eventId, func, maxTime = 500) => {
+  let eventReceived = false;
+  const eventCallback = () => {
+    eventReceived = true;
+  };
+
+  subscribe(eventId, eventCallback);
+
+  await func();
+
+  const start = Date.now();
+
+  while (Date.now() - start < maxTime) {
```

Comment:
why do a busy wait and not resolve a promise on eventCallback?

Reply:
Can you elaborate please? Not sure I understand

Reply:
If I don't do busy wait, I think the test might finish before getting the promise resolved?

Reply:
Something like that:
```
const execAndWaitForRequest = async (eventId, func, maxTime = 500) => {
  let eventPromise;
  const eventCallback = () => {
    eventPromise.resolve(`The event ${eventId} was sent to Wilco`);
  };

  const subscribePromise = new Promise(resolve => {
    eventPromise = resolve;
    subscribe(eventId, eventCallback);
  });

  try {
    await Promise.race([
      new Promise(resolve => setTimeout(resolve, maxTime)),
      subscribePromise
    ]);
  } catch {
    unsubscribe(eventId, eventCallback);
    throw new Error(`The event ${eventId} was not sent to Wilco`);
  } finally {
    unsubscribe(eventId, eventCallback);
  }
};
```

Reply:
I like Shem's solution. @NivRichter If you decide to use it, change also for the backend tests

Reply:
Done